pkg/bloomgateway/bloomgateway.go

```diff
@@ -355,19 +312,250 @@ func (g *Gateway) FilterChunkRefs(ctx context.Context, req *logproto.FilterChunk
+		}
+		idx = newIdx
+
+		if len(items) == 0 {
+			continue
+		}
+		w.metrics.dequeuedTasks.WithLabelValues(w.id).Add(float64(len(items)))
+
+		tasksPerDay := make(map[time.Time][]Task)
+
+		for _, item := range items {
+			task, ok := item.(Task)
+			if !ok {
+				// This really should never happen, because only the bloom gateway itself can enqueue tasks.
+				return errors.Errorf("failed to cast dequeued item to Task: %v", item)
+			}
+			level.Debug(w.logger).Log("msg", "dequeued task", "task", task.ID)
+			w.tasks.Delete(task.ID)
+
+			fromDay, throughDay := task.Bounds()
+
+			if fromDay.Equal(throughDay) {
+				tasksPerDay[fromDay] = append(tasksPerDay[fromDay], task)
+			} else {
+				// split task into separate tasks per day
+				for i := fromDay; i.Before(throughDay); i = i.Add(24 * time.Hour) {
+					r := filterRequestForDay(task.Request, i)
+					t := task.WithRequest(r)
+					tasksPerDay[i] = append(tasksPerDay[i], t)
+				}
+			}
+		}
+
+		for day, tasks := range tasksPerDay {
+			logger := log.With(w.logger, "day", day)
+			level.Debug(logger).Log("msg", "process tasks", "tasks", len(tasks))
+
+			it := newTaskMergeIterator(tasks...)
+
+			fingerprints = fingerprints[:0]
+			for it.Next() {
+				// fingerprints are already sorted. we can skip duplicates by checking
+				// if the next is greater than the previous
+				fp := uint64(it.At().Fp)
+				if len(fingerprints) > 0 && fp <= fingerprints[len(fingerprints)-1] {
+					continue
+				}
+				fingerprints = append(fingerprints, fp)
+			}
+
+			it.Reset()
+
+			storeFetchStart := time.Now()
+			// GetBlockQueriers() waits until all blocks are downloaded and available for querying.
+			// TODO(chaudum): Add API that allows to process blocks as soon as they become available.
+			// This will require to change the taskMergeIterator to a slice of requests so we can seek
+			// to the appropriate fingerprint range within the slice that matches the block's fingerprint range.
+			bqs, err := w.store.GetBlockQueriers(taskCtx, tasks[0].Tenant, day, day.Add(24*time.Hour).Add(-1*time.Nanosecond), fingerprints)
+			w.metrics.storeAccessLatency.WithLabelValues(w.id, "GetBlockQueriers").Observe(time.Since(storeFetchStart).Seconds())
+			if err != nil {
+				for _, t := range tasks {
+					t.ErrCh <- err
+				}
+				// continue with tasks of next day
+				continue
+			}
+
+			// No blocks found.
+			// Since there are no blocks for the given tasks, we need to return the
+			// unfiltered list of chunk refs.
+			if len(bqs) == 0 {
+				level.Warn(logger).Log("msg", "no blocks found")
+				for _, t := range tasks {
+					for _, ref := range t.Request.Refs {
+						t.ResCh <- v1.Output{
+							Fp:   model.Fingerprint(ref.Fingerprint),
+							Chks: convertToChunkRefs(ref.Refs),
+						}
+					}
+				}
+				// continue with tasks of next day
+				continue
+			}
+
+			hasNext := it.Next()
+			for _, bq := range bqs {
+				requests = requests[:0]
+				for hasNext && it.At().Fp <= bq.MaxFp {
```

Comment:
nit: I think this would better read as:
```go
for it.Next() && it.At().Fp <= bq.MaxFp {
...
}
```

Reply:
It would be better to read, but it would not work as you think:

When you call `it.Next()` the iterator will proceed to the next item. However, if the condition `it.At().Fp <= bq.MaxFp` does not match, the loop is exited.
Then the loop is started for the next `bq` again, and first, `it.Next()` will be called, proceeding to the next item, and therefore skipping one item.

Reply:
Oh I see. Thanks for the clarification. I'd probably add a comment explaining that.